File: src/main/java/com/google/android/finsky/api/ProtoDfeRequest.java
package com.google.android.finsky.api;

import com.android.volley.Response;
import com.google.android.finsky.protos.nano.Messages;
import com.google.protobuf.nano.MessageNano;

class ProtoDfeRequest extends DfeRequest
{
    private final MessageNano mRequest;
    
    ProtoDfeRequest(final String s, final MessageNano mRequest, final DfeApiContext dfeApiContext, final Response.Listener<Messages.Response.ResponseWrapper> listener, final Response.ErrorListener errorListener) {
        super(1, s, dfeApiContext, listener, errorListener);
        this.mRequest = mRequest;
        this.setShouldCache(false);
    }
    
    @Override
    public byte[] getBody() {
        return MessageNano.toByteArray(this.mRequest);
    }
    
    @Override
    public String getBodyContentType() {
        return "application/x-protobuf";
    }
}

File: playstore/src/main/java/finsky/api/ProtoDfeRequest.kt
package finsky.api

import com.android.volley.Response
import com.google.protobuf.MessageLite
import finsky.protos.Messages

internal open class ProtoDfeRequest(
        url: String,
        val request: MessageLite,
        dfeApiContext: DfeApiContext,
        listener: Response.Listener<Messages.Response.ResponseWrapper>,
        errorListener: Response.ErrorListener) : DfeRequest(Method.POST, url, dfeApiContext, listener, errorListener) {

    init {
        this.setShouldCache(false)
    }

    override fun getBody(): ByteArray {
        return request.toByteArray()
    }

    override fun getBodyContentType(): String {
        return "application/x-protobuf"
    }
}